source/darwin/securityframework_rsa.c

```diff
@@ -0,0 +1,545 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/cal/private/rsa.h>
+
+#include <aws/cal/cal.h>
+#include <aws/common/encoding.h>
+
+#include <Security/SecKey.h>
+#include <Security/Security.h>
+
+struct sec_rsa_key_pair {
+    struct aws_rsa_key_pair base;
+    CFAllocatorRef cf_allocator;
+    SecKeyRef priv_key_ref;
+    SecKeyRef pub_key_ref;
+};
+
+static void s_rsa_destroy_key(struct aws_rsa_key_pair *key_pair) {
+    if (key_pair == NULL) {
+        return;
+    }
+
+    struct sec_rsa_key_pair *rsa_key = key_pair->impl;
+
+    if (rsa_key->pub_key_ref) {
+        CFRelease(rsa_key->pub_key_ref);
+    }
+
+    if (rsa_key->priv_key_ref) {
+        CFRelease(rsa_key->priv_key_ref);
+    }
+
+    if (rsa_key->cf_allocator) {
+        aws_wrapped_cf_allocator_destroy(rsa_key->cf_allocator);
+    }
+
+    aws_mem_release(key_pair->allocator, rsa_key);
+}
+
+/*
+ * return either pointer to a key algo (global consts) or NULL in case algo is
+ * not supported.
+ */
+static int s_map_rsa_encryption_algo_to_sec(enum aws_rsa_encryption_algorithm algorithm, SecKeyAlgorithm *out) {
+
+    switch (algorithm) {
+        case AWS_CAL_RSA_ENCRYPTION_PKCS1_5:
+            *out = kSecKeyAlgorithmRSAEncryptionPKCS1;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_ENCRYPTION_OAEP_SHA256:
+            *out = kSecKeyAlgorithmRSAEncryptionOAEPSHA256;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_ENCRYPTION_OAEP_SHA512:
+            *out = kSecKeyAlgorithmRSAEncryptionOAEPSHA512;
+            return AWS_OP_SUCCESS;
+    }
+
+    return AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+}
+
+/*
+ * return either pointer to a key algo (global consts) or NULL in case algo is
+ * not supported.
+ */
+static int s_map_rsa_signing_algo_to_sec(enum aws_rsa_signing_algorithm algorithm, SecKeyAlgorithm *out) {
+
+    switch (algorithm) {
+        case AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256:
+            *out = kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_SIGNATURE_PSS_SHA256:
+            if (__builtin_available(macos 10.13, ios 11.0, tvos 11.0, watchos 4.0, *)) {
```

Comment:
are you referring to linking issues similar to https://github.com/curl/curl/issues/4314? 
i.e. in some cases optimizer not being able to optimize away the call to __builtin_available during static build and code having to call into dynamic lib at runtime, which might fail? 
I dont think  we try to prevent it now. 
```
#if (defined(__MAC_OS_X_VERSION_MAX_ALLOWED) && (__MAC_OS_X_VERSION_MAX_ALLOWED >= 101300 /* macOS 10.13 */)) ||       \
(defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= 110000 /* iOS v11 */))
```
just checks that the target os supports the symbol we want at compile time. 
and then __builtin_available does runtime check that the symbol is available on the system. 
So the current logic is basically if we are targeting a really old mac version, dont include symbol at compile time at all and if we are building for a newer system, try to do a runtime check. 

Ill add the compile time check.

Reply:
While we are on this topic - our versioning checks on mac are not completely correct. For aes we are checking for the usage of 10.13 functions and falling back to older ones. The compile time check we added last time to aes was there to prevent a warning from clang about function unavailability when setting target mac os to 10.9. 
Problem is there are 2 separate unavailability warnings - one for 10.13+ functions used on older targets and that warning is enabled by default, and another warning for <10.13 functions and that one is disable by default. And we do use a lot of 10.12+ functions that are unguarded right now (ecc is primary offender for that, and rsa will have the same). So effectively c-cal cannot be used on anything older than 10.12 right now (it will probably fail to load at runtime).
We should discuss locking down min version to 10.12 instead of current 10.9.